### DISCUSSION TITLE
I can't compile qsv from source...

### DISCUSSION THREAD

**Discussion by jqnatividad** [dathere/qsv#335] [FAQ]
## I tried building qsv, but nothing happened, and no binaries were created!

Doing a `cargo build --release` is not enough. You need to specify which qsv variant you want to build:

```
# for qsv
cargo build --release --locked --features apply,luau,fetch,foreach,geocode,polars,python,to,to_parquet,feature_capable
# or shorthand
cargo build --release --locked -F all_features

# for qsvlite
cargo build --release --locked --features lite
# for qsvdp
cargo build --release --locked --features datapusher_plus,luau
```
## I can't build qsv with the `python` feature.

If you get the error:
```
error: linking with `cc` failed: exit status: 1
...
  = note: /usr/bin/ld: cannot find -lpython3.12
          collect2: error: ld returned 1 exit status
```

You need the python 3.12 development libraries to build qsv with the python feature. On Ubuntu Linux:

`sudo apt-get install python312-dev`

## I need to use qsv on an old Linux distro which doesn't have the latest GNU libraries.

You can build qsv and statically link it with [musl](https://musl.libc.org/), so you can run it on older distros.  To do so, you'll need to install `mustl-tools` as well:

```
$ sudo apt-get install musl-tools
$ cargo build --release --features apply,luau,fetch,foreach,geocode,polars,python,to,to_parquet,feature_capable --target x86_64-unknown-linux-musl
```

**Comment by derekmahar** [comment]
How might I build qsv using a version of Rust that differs from that which `rust-version` specifies in `Cargo.toml`?  For example, how might I build qsv 0.71.0 using Rust 1.64 instead of Rust 1.65?

**Reply by jqnatividad** [reply]
@derekmahar , for 0.71.0 specifically, it should just be a simple matter of changing rust-version to 1.64.

Be aware however that in general, I try to use new language features as soon as they are stabilized, if it makes sense for qsv - primarily, from a performance and maintainability perspective.

However, for 1.65, I've yet to find a use for its headline features - GAT and `let else`, so you should be able to compile using 1.64 by changing rust-version in cargo.toml.

**Reply by derekmahar** [reply]
Thank you.  I will compile qsv by changing `rust-version` to 1.64 or wait until Rust 1.65 is available in the Nix repository.
